zipkin-storage/elasticsearch-http/src/main/java/zipkin/storage/elasticsearch/http/ElasticsearchHttpSpanConsumer.java

```diff
@@ -37,29 +38,95 @@
 
   final ElasticsearchHttpStorage es;
   final IndexNameFormatter indexNameFormatter;
+  private final static int MAX_CACHE_DAYS = 2;
+  private final static int MAX_CACHE_ENTRIES = 1000;
+
+  private final static Map<String, Set<Pair<String>>> indexToServiceSpansCache = new LinkedHashMap<String, Set<Pair<String>>>() {
+    @Override
+    protected boolean removeEldestEntry(Map.Entry<String, Set<Pair<String>>> eldest) {
+      return size() > MAX_CACHE_DAYS;
+    }
+  };
+  private Map<String, Set<Pair<String>>> indexToServiceSpans = new LinkedHashMap<>();
 
   ElasticsearchHttpSpanConsumer(ElasticsearchHttpStorage es) {
     this.es = es;
     this.indexNameFormatter = es.indexNameFormatter();
   }
 
   @Override public void accept(List<Span> spans, Callback<Void> callback) {
+    Callback<Void> callbackWrapper = new Callback<Void>() {
+        @Override
+        public void onSuccess(Void value) {
+            callback.onSuccess(value);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            clearIndexToServiceSpansCache(indexToServiceSpans, indexToServiceSpansCache);
+            callback.onError(t);
+        }
+    };
+
     if (spans.isEmpty()) {
-      callback.onSuccess(null);
+      callbackWrapper.onSuccess(null);
       return;
     }
     try {
       HttpBulkIndexer indexer = new HttpBulkIndexer("index-span", es);
-      Map<String, Set<Pair<String>>> indexToServiceSpans = indexSpans(indexer, spans);
+      indexToServiceSpans = indexSpans(indexer, spans);
+
+      // remove already cached items from indexToServiceSpans, add new items to cache
+      for (Map.Entry<String, Set<Pair<String>>> entry : indexToServiceSpans.entrySet()) {
+        Set<Pair<String>> serviceSpansEntryCached = null;
+        synchronized(indexToServiceSpansCache) {
+          serviceSpansEntryCached = indexToServiceSpansCache.get(entry.getKey());
+        }
+        if (serviceSpansEntryCached == null) {
+          synchronized(indexToServiceSpansCache) {
```

Comment:
next step is to make some concurrency tests... something like https://github.com/openzipkin/brave/blob/7fd9bae7ee3ba8abaa536ad1d29634099a337d83/brave/src/test/java/brave/CurrentTracingTest.java#L53

you'll want to make sure that they break when you remove the synchronized block